### PR TITLE
fix: double_count default is no double count 

### PR DESCRIPTION
**Pull request by mpark63**
## double_count redefined
The default value (or the absence) of `double_count` signified that the distribution could overlap freely with any other. We did this because it saved us typing. To avoid confusion, however, we are changing the logic so that the absence of `double_count` indicates no double count.

## Implementation 
- modified double_count field checking for distributions and fine requirements 
Screen Shot 2022-07-27 at 8 44 27 PM
- removed all instances of `double_count: ['N/A']` in majors.tsx
- add `double_count: ['All']` in majors.tsx where necessary 

## double_count guide 
- undefined or []: cannot double count with any (aka is exclusive) 
- ['A', 'B']: can double count with A or B 
- ['All']: can overlap with all distributions 
    - saves us from typing all the distributions 

## testing 
Tested locally. Other open PRs will need to pull these changes and ensure their code is compatible with the new `double_count`.

### CODE REVIEW

File: lib/components/dashboard/degree-info/DistributionBarsJSX.tsx [hunk 1]
```diff
@@ -199,25 +199,29 @@ const DistributionBarsJSX: FC<{ major: Major }> = ({ major }) => {
     // If double_count is string[], the specified distributions / fine requirements are 'whitelisted'
     // When a course satisfies a distribution, double_count takes the value of distribution.double_count
     if (!courseObj) return;
-    let distDoubleCount: string[] | undefined = undefined; // initial value
+    let distDoubleCount: string[] | undefined = ['All']; // double count by defualt
     reqs.forEach((reqGroup, i) => {
       let req = reqGroup[1][0]; // general distribution
       if (
-        (!distDoubleCount || distDoubleCount.includes(req.name)) &&
+        distDoubleCount &&
```
Comment (mattliu-mygit) [inline lib/components/dashboard/degree-info/DistributionBarsJSX.tsx:206]: distDoubleCount will never be undefined, false, 0, an empty string, or null. Thus, it will always compute to true and this check is extraneous and unnecessary.

File: lib/components/dashboard/degree-info/DistributionBarsJSX.tsx [hunk 2]
```diff
@@ -199,25 +199,29 @@ const DistributionBarsJSX: FC<{ major: Major }> = ({ major }) => {
...
-        let fineDoubleCount: string[] | undefined = undefined;
+        let fineDoubleCount: string[] | undefined = ['All'];
         reqGroup[1].forEach((req: requirements, j: number) => {
           // fine reqs
           if (j !== 0) {
             // 0 is the general distribution, not fine req
             let fineReq = reqGroup[1][j];
             if (
-              (!fineDoubleCount || fineDoubleCount.includes(fineReq.name)) &&
+              fineDoubleCount &&
```
Comment (mattliu-mygit) [inline lib/components/dashboard/degree-info/DistributionBarsJSX.tsx:222]: similar comment about a similar line above

File: lib/components/dashboard/degree-info/DistributionBarsJSX.tsx [hunk 3]
```diff
@@ -199,25 +199,29 @@ const DistributionBarsJSX: FC<{ major: Major }> = ({ major }) => {
...
         (req.fulfilled_credits < req.required_credits ||
           (req.required_credits === 0 && req.fulfilled_credits === 0)) &&
         checkRequirementSatisfied(req, courseObj)
       ) {
         reqs[i][1][0].fulfilled_credits += parseInt(courseObj.credits);
         distDoubleCount = req.double_count; // set double_count, if any
-        let fineDoubleCount: string[] | undefined = undefined;
+        let fineDoubleCount: string[] | undefined = ['All'];
         reqGroup[1].forEach((req: requirements, j: number) => {
           // fine reqs
```
Comment (mattliu-mygit) [inline lib/components/dashboard/degree-info/DistributionBarsJSX.tsx:217]: The fine reqs nesting is a bit hard to follow. Can you add comments explaining what's happening and also take finereqs out of the nesting (flatten) or take it out as a helper method (unflattened)?

File: lib/resources/majors.tsx [hunk 4]
```diff
@@ -173,34 +178,39 @@ const baCogSci: Major = {
             'At least 2 courses must be at the 300 level or above.',
           required_credits: 12,
           criteria: 'COGS-COGPSY[T]',
+          double_count: ['All'],
```
Comment (mattliu-mygit) [inline lib/resources/majors.tsx:181]: On the Major schema in the backend, lets make this the default value for double_count for Major documents.

File: lib/components/dashboard/degree-info/DistributionBarsJSX.tsx (hunk 1 above)
Comment (mattliu-mygit) [reply lib/components/dashboard/degree-info/DistributionBarsJSX.tsx:206]: Or could it be undefined (ie. nonexistent). This makes sense if it could be nonexistent, but I think an empty array makes more sense as it's already checked by "distDoubleCount.includes(req.name)", saving one computation and cleaning up the logic. 

Basically, we make double_count required, and empty arrays means it can't double count with anything else.

File: lib/resources/majors.tsx [hunk 5]
```diff
@@ -868,7 +881,6 @@ const bsBME: Major = {
       description:
         'Students are required to take at least one semester of programming from a select set of gateway computing courses.',
       criteria: 'EN.500.112[C]^OR^EN.500.113[C]^OR^EN.500.114[C]',
-      double_count: ['N/A'],
```
Comment (mattliu-mygit) [inline lib/resources/majors.tsx:871]: Instead of removing this, can we make it an empty array?

**Review by mattliu-mygit — CHANGES_REQUESTED** [review]
Please view comments ^.^ good work!

File: lib/components/dashboard/degree-info/DistributionBarsJSX.tsx (hunk 1 above)
Comment (mpark63) [reply lib/components/dashboard/degree-info/DistributionBarsJSX.tsx:206]: I think `distDoubleCount` could be undefined if the distribution does not have the field. As you said, it's optional atm.

logging `distDoubleCount`: 
Screen Shot 2022-07-28 at 2 07 28 PM

File: lib/components/dashboard/degree-info/DistributionBarsJSX.tsx (hunk 3 above)
Comment (mpark63) [reply lib/components/dashboard/degree-info/DistributionBarsJSX.tsx:217]: good idea, taken out as helper method

File: lib/resources/majors.tsx (hunk 4 above)
Comment (mpark63) [reply lib/resources/majors.tsx:181]: in backend yes

File: lib/resources/majors.tsx (hunk 5 above)
Comment (mpark63) [reply lib/resources/majors.tsx:871]: we could, but I prefer having !double_count as no double_count. Less typing :/

Comment (mattliu-mygit) [reply lib/resources/majors.tsx:871]: Resolved offline.

**Review by mattliu-mygit — APPROVED** [review]

File: lib/components/dashboard/degree-info/DistributionBarsJSX.tsx (hunk 1 above)
Comment (mattliu-mygit) [reply lib/components/dashboard/degree-info/DistributionBarsJSX.tsx:207]: agreed. resolved.
